[Merged by Bors] - feat(Combinatorics/Enumerative/Stirling.lean): define Stirling numbers of the first and second kind

BeibeiX0 opened the pull request:
This PR introduces the unsigned Stirling numbers of the first kind (`Nat.stirlingFirst`)
and the Stirling numbers of the second kind (`Nat.stirlingSecond`), along with
their defining recurrence relations and several fundamental properties.

Main definitions:

* `Nat.stirlingFirst n k`: counts the number of permutations of `n` elements with `k` disjoint cycles.
* `Nat.stirlingSecond n k`: counts the number of ways to partition `n` elements into `k` non-empty subsets.

Main results include:

* Recurrence relations for both kinds.
* Vanishing conditions when `n < k`.
* Diagonal values like `stirlingFirst n n = 1`, `stirlingSecond n n = 1`.
* Identities such as `stirlingFirst (n + 1) 1 = n!` and
  `stirlingSecond (n + 1) 1 = 1`.

References:

* [Stirling numbers of the first kind – Wikipedia](https://en.wikipedia.org/wiki/Stirling_numbers_of_the_first_kind)
*  [Stirling numbers of the second kind – Wikipedia](https://en.wikipedia.org/wiki/Stirling_numbers_of_the_second_kind)

Co-authored-by: Beibei Xiong <xbbng.0@gmail.com>
Co-authored-by: Shao Yu <18217703912@163.com>
Co-authored-by: Weijie Jiang <xin82436@gmail.com>

---

BeibeiX0 committed:
initial commit for Stirling numbers

BeibeiX0 committed:
feat(Combinatorics/Enumerative/Stirling): define Stirling numbers of both kinds

BeibeiX0 committed:
feat(Combinatorics/Enumerative/Stirling): define Stirling numbers of both kinds

BeibeiX0 committed:
fix: adjust simp attributes to avoid CI failure

BeibeiX0 committed:
Fix:delete batteries

kbuzzard commented:
Can you please remove the "batteries" file you added?

BeibeiX0 committed:
Remove accidentally added batteries submodule

BeibeiX0 commented:
> Can you please remove the "batteries" submodule which you (accidentally?) added?

Thanks for catching that! The `batteries` submodule has been removed.

BeibeiX0 committed:
rename definitions to stirlingFirst and stirlingSecond"

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,203 @@
+theorem stirlingFirst_succ_zero (n : ℕ) : stirlingFirst (succ n) 0 = 0 := by
+  simp [stirlingFirst]
+
+theorem stirlingFirst_succ_left (n k : ℕ) (hk : 0 < k) :
```
```suggestion
theorem stirlingFirst_succ_left (n k : ℕ) (hk : k ≠ 0) :
```
Mathlib style is to write "not equal to zero" for naturals, rather than "positive", in assumptions: it should be fairly easy to adjust the proof here to accommodate. And similarly below!

BeibeiX0 replied:
---

Thank you for the suggestion! I understand that Mathlib generally prefers using `≠ 0` over `0 < n` when referring to natural numbers in assumptions.

That said, I was wondering whether both styles are acceptable in context. For instance, in `Mathlib/Data/Nat/Choose/Basic.lean`, the theorem `choose_succ_left` is stated using `0 < k`:

```lean
theorem choose_succ_left (n k : ℕ) (hk : 0 < k) :
  choose (n + 1) k = choose n (k - 1) + choose n k := by
  obtain ⟨l, rfl⟩ : ∃ l, k = l + 1 := Nat.exists_eq_add_of_le' hk
  rfl
```

So I’m curious if using `0 < k` is also considered stylistically acceptable in such settings, or if I should change it to `k ≠ 0` here as well. Looking forward to your thoughts!

b-mehta replied:
I would say that those should be changed too, but that can be another PR.

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,203 @@
+    stirlingFirst (n + 1) (k + 1) = n * stirlingFirst n (k + 1) +  stirlingFirst n k := by
+  rw [stirlingFirst]
+
+theorem stirlingFirst_eq_zero_of_lt : ∀ {n k : ℕ}, n < k → stirlingFirst n k= 0
```
```suggestion
theorem stirlingFirst_eq_zero_of_lt : ∀ {n k : ℕ}, n < k → stirlingFirst n k = 0
```

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,203 @@
+theorem stirlingSecond_self (n : ℕ) : stirlingSecond n n = 1 := by
+  induction n <;> simp [*, stirlingSecond, stirlingSecond_eq_zero_of_lt (lt_succ_self _)]
+
+theorem stirlingSecond_one_right (n : ℕ) : stirlingSecond (n+1) 1 = 1 := by
```
```suggestion
theorem stirlingSecond_one_right (n : ℕ) : stirlingSecond (n + 1) 1 = 1 := by
```

BeibeiX0 replied:
Thanks for the detailed comments! I've gone through and fixed all the spacing and casing issues accordingly. Please let me know if I missed anything.

b-mehta reviewed (approved):
This looks generally pretty good! I've left a couple of style comments, once they're addressed I'll look again but it should be pretty ready to go.

Don't forget to update the names in your PR description, and the links are currently to the signed rather than unsigned version.

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,203 @@
+# Main definitions
+
+* `Nat.stirlingFirst`: the number of ways to partition n distinct elements into k non-empty
+cycles, Defined by the recursive relationship it satisfies.
```
```suggestion
  cycles, defined by the recursive relationship it satisfies.
```

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,203 @@
+* `Nat.stirlingFirst`: the number of ways to partition n distinct elements into k non-empty
+cycles, Defined by the recursive relationship it satisfies.
+* `Nat.stirlingSecond`: the number of ways to partition n distinct elements into k non-empty
+subsets, Defined by the recursive relationship it satisfies.
```
```suggestion
  subsets, defined by the recursive relationship it satisfies.
```

eric-wieser commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,203 @@
+    exact ih
+
+theorem stirlning_second_succ_self_left (n : ℕ) :
+    stirlingSecond (n + 1) n = (n * (n + 1)) / 2 := by
```
```suggestion
    stirlingSecond (n + 1) n = (n + 1).choose 2 := by
```
perhaps?

BeibeiX0 replied:
I agree this is a better formulation. Thanks! I've made the change.

BeibeiX0 committed:
style(Combinatorics/Enumerative/Stirling): fix formatting (spaces, casing, naming)

BeibeiX0 committed:
style(Stirling): replace `0 < n` / `0 < k` with `n ≠ 0` / `k ≠ 0` for naturals

BeibeiX0 committed:
chore(Stirling): fix author name

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,210 @@
+
+namespace Nat
+
+
```
```suggestion
```

zhengfengATecnu replied:
Thanks for the review! I've pushed the changes, which include:

Fixing all formatting issues (spacing, indentation, etc.).

Refactoring the proofs to remove non-terminal simps and make them more declarative.

Updating the module docstring to the standard format.

Please let me know if there's anything else!

BeibeiX0 replied:
Thanks for the review, @b-mehta !

I've pushed the changes, which include:

- Fixing all formatting issues (spacing, indentation, etc.).

- Refactoring the proofs to remove non-terminal simps and make them more declarative.

- Updating the module docstring to the standard format.

Please let me know if there's anything else!

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,210 @@
+  induction' n with n ih
+  · simp [stirlingSecond]
+  · rw [stirlingSecond_zero_right']
+    simp
```
This is a non-terminal simp, which should be avoided in mathlib, see https://leanprover-community.github.io/extras/simp.html#non-terminal-simps
There are a couple like this in the file, please make sure to adjust.

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,210 @@
+    have hnk : n < k := Nat.lt_of_succ_lt_succ hk
+    have hnk1 : n < k + 1 := Nat.lt_of_succ_lt hk
+    rw [stirlingSecond_succ_succ, stirlingSecond_eq_zero_of_lt hnk,
+     stirlingSecond_eq_zero_of_lt hnk1]
```
```suggestion
      stirlingSecond_eq_zero_of_lt hnk1]
```

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,210 @@
+
+theorem stirlingSecond_succ_succ (n k : ℕ) :
+    stirlingSecond (n + 1) (k + 1) =
+     (k + 1) * stirlingSecond n (k + 1) + stirlingSecond n k := by
```
```suggestion
       (k + 1) * stirlingSecond n (k + 1) + stirlingSecond n k := by
```

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,210 @@
+  rfl
+
+theorem stirlingFirst_succ_right (n k : ℕ) (hn : n ≠ 0) :
+    stirlingFirst n (k + 1)
+     = (n - 1) * stirlingFirst (n - 1) (k + 1) + stirlingFirst (n - 1) k := by
```
```suggestion
    stirlingFirst n (k + 1) =
      (n - 1) * stirlingFirst (n - 1) (k + 1) + stirlingFirst (n - 1) k := by
```

BeibeiX0 committed:
refactor(Combinatorics/Enumerative/Stirling): align with style guide

BeibeiX0 committed:
fixchore(Stirling): remove unused import `Mathlib.Data.Nat.ModEq`

BeibeiX0 committed:
Fix non-terminal simps in Stirling number proofs

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,170 @@
+  induction n <;> simp [*, stirlingSecond, stirlingSecond_eq_zero_of_lt (lt_succ_self _)]
+
+theorem stirlingSecond_one_right (n : ℕ) : stirlingSecond (n + 1) 1 = 1 := by
+  simp [stirlingSecond]
```
This one is still a non-terminal simp

BeibeiX0 replied:
Apologies for the oversight, @b-mehta ! It seems my previous fix was incomplete.
I've pushed a new commit that should now correctly resolve all instances of non-terminal simps across the file. 
Thank you for your patience and for catching that.

b-mehta commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,170 @@
+
+# References:
+
+* [Stirling Numbers of the First Kind – MathWorld](https://mathworld.wolfram.com/StirlingNumberoftheFirstKind.html)
+* [Stirling Numbers of the Second Kind – MathWorld](https://mathworld.wolfram.com/StirlingNumberoftheSecondKind.html)
```
These links are for the signed Stirling numbers, not the unsigned ones which you have in this file

BeibeiX0 replied:
Thanks for catching that!
I've updated the links to point to the correct pages for the unsigned Stirling numbers of the first kind and the Stirling numbers of the second kind, ensuring they match the definitions in this file.

BeibeiX0 committed:
refactor: Fix non-terminal simps in Stirling number proofs

BeibeiX0 committed:
docs: Fix incorrect reference links for Stirling numbers

BeibeiX0 committed:
refactor: Fix non-terminal simps in Stirling number proofs

BeibeiX0 committed:
delete unused tactic

BeibeiX0 committed:
chore(Stirling): improve tactic clarity and proof readability

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+
+# Main definitions
+
+* `Nat.stirlingFirst`: the number of ways to partition n distinct elements into k non-empty
+  cycles, defined by the recursive relationship it satisfies.
```
```suggestion
* `Nat.stirlingFirst`: the number of ways to partition `n` distinct elements into `k` non-empty
  cycles, defined by the recursive relationship it satisfies.`
```

BeibeiX0 replied:
Thank you for your careful inspection. I’ve added the missing backticks in the docstring.

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+
+* `Nat.stirlingFirst`: the number of ways to partition n distinct elements into k non-empty
+  cycles, defined by the recursive relationship it satisfies.
+* `Nat.stirlingSecond`: the number of ways to partition n distinct elements into k non-empty
+  subsets, defined by the recursive relationship it satisfies.
```
```suggestion
* `Nat.stirlingSecond`: the number of ways to partition `n` distinct elements into `k` non-empty
  subsets, defined by the recursive relationship it satisfies.
```

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+
+# References:
+
+* [Stirling Numbers of the First Kind – ProofWiki](https://proofwiki.org/wiki/Definition:Stirling_Numbers_of_the_First_Kind/Unsigned)
+* [Stirling Numbers of the Second Kind – ProofWiki](https://proofwiki.org/wiki/Definition:Stirling_Numbers_of_the_Second_Kind)
```
I would rather add a standard reference such as Knuth, The Art of Computer Programming, volume 1, §1.2.

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+
+theorem stirlingFirst_succ_left (n k : ℕ) (hk : k ≠ 0) :
+    stirlingFirst (n + 1) k = n * stirlingFirst n k + stirlingFirst n (k - 1) := by
+  have hk : 0 < k := Nat.pos_of_ne_zero hk
+  obtain ⟨l, rfl⟩ : ∃ l, k = l + 1 := Nat.exists_eq_add_of_le' hk
```
```suggestion
  obtain ⟨l, rfl⟩ := Nat.exists_eq_add_of_le' (Nat.pos_of_ne_zero hk)
```

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+theorem stirlingFirst_succ_right (n k : ℕ) (hn : n ≠ 0) :
+    stirlingFirst n (k + 1) =
+      (n - 1) * stirlingFirst (n - 1) (k + 1) + stirlingFirst (n - 1) k := by
+  have hn : 0 < n := Nat.pos_of_ne_zero hn
+  obtain ⟨l, rfl⟩ : ∃ l, n = l + 1 := Nat.exists_eq_add_of_le' hn
```
```suggestion
  obtain ⟨l, rfl⟩ := Nat.exists_eq_add_of_le' (Nat.pos_of_ne_zero hn)
```

BeibeiX0 replied:
Thanks for the suggestion! I've applied the change.

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+  | _, 0, hk => absurd hk (Nat.not_lt_zero _)
+  | 0, _ + 1, _ => by rw [stirlingFirst]
+  | n + 1, k + 1, hk => by
+    have hnk : n < k := Nat.lt_of_succ_lt_succ hk
```
You can probably avoid the `have` by inserting them into the command.

BeibeiX0 replied:
Thank you! I’ve removed all intermediate haves and inlined the reasoning directly into the rw statements as suggested.

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+
+theorem stirlingSecond_succ_left (n k : ℕ) (hk : k ≠ 0) :
+    stirlingSecond (n + 1) k = k * stirlingSecond n k + stirlingSecond n (k - 1) := by
+  have hk : 0 < k := Nat.pos_of_ne_zero hk
```
Same comment as above.

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+theorem stirlingSecond_succ_right (n k : ℕ) (hn : n ≠ 0) :
+    stirlingSecond n (k + 1) =
+      (k + 1) * stirlingSecond (n - 1) (k + 1) + stirlingSecond (n - 1) k := by
+  have hn : 0 < n := Nat.pos_of_ne_zero hn
```
idem.

AntoineChambert-Loir commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+  | _, 0, hk => absurd hk (Nat.not_lt_zero _)
+  | 0, _ + 1, _ => by rw [stirlingSecond]
+  | n + 1, k + 1, hk => by
+    have hnk : n < k := Nat.lt_of_succ_lt_succ hk
```
idem

BeibeiX0 committed:
fix(Combinatorics.Stirling): add missing colons in comments; inline have

BeibeiX0 commented on Mathlib/Combinatorics/Enumerative/Stirling.lean:
```diff
@@ -0,0 +1,176 @@
+# The Stirling numbers of the first kind
+
+The unsigned Stirling numbers of the first kind, represent the number of ways
+to partition n distinct elements into k non-empty cycles.
```
Thanks for pointing that out! I’ve added the missing backticks in the docstring.

BeibeiX0 commented:
Thanks for your review and comments. I’ve addressed the issues and
submitted the updated version.😀

Bhavik Mehta ***@***.***> 于2025年7月21日周一 03:11写道：

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Mathlib/Combinatorics/Enumerative/Stirling.lean
> <https://github.com/leanprover-community/mathlib4/pull/26290#discussion_r2217923100>
> :
>
> > +* `Nat.stirlingFirst`: the number of ways to partition n distinct elements into k non-empty
> +  cycles, defined by the recursive relationship it satisfies.
>
> I think the ` at the end of the sentence might have been accidental here?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/leanprover-community/mathlib4/pull/26290#discussion_r2217923100>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BFDDMGE2FPPK73HT2WIW3ND3JPSVZAVCNFSM6AAAAAB74QCO4WVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMZWGEZTCMZQGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

BeibeiX0 committed:
chore(authors): update authors list

BeibeiX0 commented:
Hi @eric-wieser , I’ve made the changes based on the other reviewers’ feedback. Could you please review it when you have a chance? Thanks!

b-mehta reviewed (commented):
Thanks!

bors merge

BeibeiX0 committed:
git commit -m "Fix formatting: add missing space"

fpvandoorn commented:
For future reference: within PRs we prefer if you don't force-push after having received reviews. 
If you're used to rebasing master, please just merge master instead.

bors merge
